### PR TITLE
WIP: Add ability to disregard max file size for specific files

### PR DESCRIPTION
**Pull request by ijsnow**
# Add ability to disregard max file size for specific files

This PR adds a `large_files` flag that is a comma separated list of files that we want to index regardless of the file size.

This is part of the solution for https://github.com/sourcegraph/sourcegraph/issues/1624.

## Testing plan

1. Make a large file. I ran this program:

```go
package main

import (
	"fmt"
	"log"
	"os"
)

func main() {
	file, err := os.Create("/tmp/large-file.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	maxSize := 20 * 1024 * 1024
	size := 0

	for size < maxSize+1 {
		n, _ := file.WriteString("test\n")
		size += n
	}

	info, _ := file.Stat()
	fmt.Printf("File size: %v\n", info.Size())
}
```

2. Create a temporary git directory and add the large file you just created

```
mkdir /tmp/big-index
cd /tmp/big-index
git init
mv /tmp/large-file.txt ./
git add .
git commit -m "add a large file"
```

3. Create an archive of the repo

```
git archive master --format=tar --output=/tmp/big-index.tar
```

4. Run the archive command

```
go run ./cmd/zoekt-archive-index -large_files large-file.txt -branch master -index /tmp/test-indexes -name big-index /tmp/big-index.tar
```

5. Make sure the contents of the big file are indexed

```
cat /tmp/test-indexes/big-index_v15.00000.zoekt
```

### CODE REVIEW

**Comment by emidoots** [comment]
Given https://github.com/sourcegraph/sourcegraph/issues/1624#issuecomment-465443996 won't you need a way to change this at runtime? Or how do you plan to update this flag when the site config changes?

**Comment by beyang** [comment]
Echoing what Stephen said, I'd be curious to understand how this arg will be set. Also, will it solve the end use case described in https://github.com/sourcegraph/sourcegraph/issues/1624 if a site admin has to explicitly list all the large files that should be indexed?

**Comment by keegancsmith** [comment]
Given the comments above it isn't clear why this is part of the solution. Could you update the linked issue with a proposal of the path you are taking to solve the issue?

**Comment by ijsnow** [comment]
@keegancsmith I'm going with @sqs's proposal https://github.com/sourcegraph/sourcegraph/issues/1624#issuecomment-465443996. Without the large files flag, how can we allow the indexing of just certain large files based using just the SizeMax option?

**Comment by ijsnow** [comment]
Changed PR to WIP in title because I haven't accounted for updating at runtime and don't see a way to switch the PR to a draft PR.

**Comment by ijsnow** [comment]
@slimsag @beyang my plan for updating this configuration option at runtime is to add a new special cased route [just like we do for debugging.](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/blob/cmd/zoekt-sourcegraph-indexserver/main.go#L206-209) This route will just update `Server.LargeFiles`. If I'm not mistaken, the indexer regularly kicks off indexing jobs and from then on the new large files value will be used.

**Comment by emidoots** [comment]
That makes sense. Given that, why don't we always just use that route to update the large files list? i.e. what is the purpose of the CLI flag? (maybe you have a good argument for why we would have it and I just don't know it)

**Comment by ijsnow** [comment]
@slimsag I've added the handler and removed the cli flag. Does this look good to you? Next steps in solving https://github.com/sourcegraph/sourcegraph/issues/1624 are to use this new option inside Sourcegraph.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (emidoots) [inline cmd/zoekt-sourcegraph-indexserver/main.go:271]: you may find that just passing JSON messages is more straightforward than trying to do a fully accurate REST API, i.e. instead of using form data. We do this elsewhere and imo it works quite nicely. i.e. everything is just POST with JSON bodies, get a JSON response back and you can neglect all other details of HTTP (response status code indicates a server failure, never an operational failure, etc.)

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (emidoots) [inline cmd/zoekt-sourcegraph-indexserver/main.go:258]: If you are going for full-on REST, should this be `OPTIONS`? or am I mistaken

**Review by emidoots — APPROVED** [review]
cool, yeah, lgtm but get approval from 1 more person just due to how many folks are chiming in here

File: cmd/zoekt-archive-index/main.go
Comment (beyang) [inline cmd/zoekt-archive-index/main.go:224]: Could this be encapsulated in a method on the Options struct (similar to [`shardName`](https://sourcegraph.com/github.com/sourcegraph/zoekt@395f561fba1e03ec7c8d0615a0b0b89db2f0afed/-/blob/build/builder.go#L150:19))? Method name would be something like `largeFilesSet`.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (ijsnow) [reply cmd/zoekt-sourcegraph-indexserver/main.go:271]: I went with a form body to be consistent with other data being sent to this service. See the rest of the `ServeHTTP` function for the other form data. I'm fine with changing it to JSON but would like to hear @keegancsmith 's opinion on the consistency here.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (ijsnow) [reply cmd/zoekt-sourcegraph-indexserver/main.go:258]: `OPTIONS` handlers return what method options are allowed for an endpoint. For example, if we supported the options method for this endpoint, it would just return `POST`. We browsers usually use this to make sure that the request it is about to send is supported.

Comment (ijsnow) [reply cmd/zoekt-sourcegraph-indexserver/main.go:258]: From https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS:

> The HTTP OPTIONS method is used to describe the communication options for the target resource. The client can specify a URL for the OPTIONS method, or an asterisk (*) to refer to the entire server.

I'm happy to change the name to settings rather than options to reduce confusion.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (beyang) [inline cmd/zoekt-sourcegraph-indexserver/main.go:257]: Add a docstring to describe what this does. Also, include whether the behavior is "this will trigger an update to the index with the new options" or "this updates the options, which will get used the next time indexing runs".

Taking a step back, what will the overall contract be on how long it takes for the config update to be fully applied (such that end users will see the changes reflected in all search results)? Will it be all or nothing or gradual?

**Review by beyang — APPROVED** [review]
LGTM pending the following:
- [ ] inline comments resolved
- [ ] @keegancsmith's approval

**Comment by keegancsmith** [comment]
> @keegancsmith I'm going with @sqs's proposal [sourcegraph/sourcegraph#1624 (comment)](https://github.com/sourcegraph/sourcegraph/issues/1624#issuecomment-465443996). Without the large files flag, how can we allow the indexing of just certain large files based using just the SizeMax option?

Sorry I didn't read @sqs's proposal. I just looked for a comment from you / PR description.

**Comment by ijsnow** [comment]
Closing in favor of a combination of merging in some changes up stream and others here: https://github.com/sourcegraph/zoekt/pull/3

**Pull request closed without merging**
